src/MALA.jl

```diff
@@ -19,48 +19,60 @@ struct GradientTransition{T<:Union{Vector, Real, NamedTuple}, L<:Real, G<:Union{
     gradient::G
 end
 
-transition(::MALA, model, params) = GradientTransition(model, params)
-
-# Store the new draw, its log density and its gradient
-GradientTransition(model::DensityModel, params) = GradientTransition(params, logdensity_and_gradient(model, params)...)
+logdensity(model::DensityModel, t::GradientTransition) = t.lp
 
 propose(rng::Random.AbstractRNG, ::MALA, model) = error("please specify initial parameters")
+function transition(sampler::MALA, model::DensityModel, params)
+    return GradientTransition(params, logdensity_and_gradient(model, params)...)
+end
 
-function propose(
+function AbstractMCMC.step(
     rng::Random.AbstractRNG,
-    spl::MALA{<:Proposal},
     model::DensityModel,
-    params_prev::GradientTransition
-)
-    proposal = propose(rng, spl.proposal(params_prev.gradient), model, params_prev.params)
-    return GradientTransition(model, proposal)
-end
-
-
-function q(
-    spl::MALA{<:Proposal},
-    t::GradientTransition,
-    t_cond::GradientTransition
-)
-    return q(spl.proposal(-t_cond.gradient), t.params, t_cond.params)
-end
-
-function logratio_proposal_density(
-    sampler::MALA{<:Proposal}, state::GradientTransition, candidate::GradientTransition
+    sampler::MALA,
+    transition_prev::GradientTransition;
+    kwargs...
 )
-    return q(sampler, state, candidate) - q(sampler, candidate, state)
+    # Extract value and gradient of the log density of the current state.
+    state = transition_prev.params
+    logdensity_state = transition_prev.lp
+    gradient_logdensity_state = transition_prev.gradient
+
+    # Generate a new proposal.
+    proposal = sampler.proposal
+    candidate = propose(rng, proposal(gradient_logdensity_state), model, state)
+
+    # Compute both the value of the log density and its gradient
+    logdensity_candidate, gradient_logdensity_candidate = logdensity_and_gradient(
+        model, candidate
+    )
+
+    # Compute the log ratio of proposal densities.
+    logratio_proposal_density = q(
+        proposal(-gradient_logdensity_candidate), state, candidate
+    ) - q(proposal(-gradient_logdensity_state), candidate, state)
```

Comment:
This is not modified by this PR but I'm a bit confused why one uses `proposal(-gradient_logdensity_candidate)` here even though the algorithm uses `proposal(gradient_logdensity_candidate)` for sampling the proposal. I guess there's a simple explanation and I'm just too tired to see it right now :sleeping:

Reply:
Not sure, honestly. @briandepasquale do you know why the negative is used here?

Reply:
Maybe let's open an issue and discuss it separately since it does not change the current implementation?

Reply:
Hi all. I recall that minus sign... Let me take a look. Yes perhaps within an issue is more appropriate. I will be a little slow though - I am deep in an academic search right now and about to start interviewing, very time intensive and stressful. But hopefully won't be months, like my original PR :)